### PR TITLE
Change initRound to set rewardIndex as 0, not WAD

### PR DESCRIPTION
**Pull request by qpzm**
## Description
- Change `initRound` to initialize `rewardIndex` as 0, not WAD
- Refactor test helpers: `toTimestamp`
- Add test helpers: `stakeAndCheck`, `extendPoolAndCheck`
- Extract `tests/scenario/stake.test.ts`, `tests/extendPool.test.ts`

## Test
All passes

### CODE REVIEW

File: test/close.test.ts [hunk 1]
```diff
@@ -2,11 +2,12 @@ import { createTestActions, TestHelperActions } from './utils/helpers';
 import { BigNumber, utils } from 'ethers';
 import { waffle } from 'hardhat';
 import { expect } from 'chai';
+import moment from 'moment';
 
 import TestEnv from './types/TestEnv';
-import { RAY, SECONDSPERDAY, WAD, MAX_UINT_AMOUNT } from './utils/constants';
+import { RAY, SECONDSPERDAY, MAX_UINT_AMOUNT } from './utils/constants';
```
Comment (jaeboklee-modo) [inline test/close.test.ts]: I think we can use `uintMax` in ethers constant instead of `MAX_UINT_AMOUNT`

Comment (qpzm) [reply test/close.test.ts]: Do you mean this? It is BigNumber and `MAX_UINT_AMOUNT` is string. 
https://docs.ethers.io/v5/api/utils/constants/#constants-MaxUint256
I'm not sure which way is better 😅.

File: test/stake.test.ts [hunk 2]
```diff
@@ -55,328 +56,99 @@ describe('StakingPool.stake', () => {
...
-      );
-
-      const poolDataAfter = await getPoolData(testEnv);
-      const userDataAfter = await getUserData(testEnv, alice);
-
-      expect(poolDataAfter).to.be.equalPoolData(expectedPoolData_2);
-      expect(userDataAfter).to.be.equalUserData(expectedUserData_2);
+      const { stakeAndCheck, extendPoolAndCheck } = actions;
+      await extendPoolAndCheck(deployer, alice, newRewardPersecond, duration);
+      await stakeAndCheck(alice, stakeAmount);
```
Comment (qpzm) [inline test/stake.test.ts:134]: Tests become simpler by adding `extendPoolAndCheck` and `stakeAndCheck`. 
These fetch data before an action and asserts the expected outcome after the action.

File: test/utils/helpers.ts [hunk 3]
```diff
@@ -86,10 +99,67 @@ export const createTestActions = (testEnv: TestEnv): TestHelperActions => {
...
+      await getTimestamp(stakeTx),
+      amount
+    );
+
+    const poolDataAfter = await getPoolData();
+    const userDataAfter = await getUserData(wallet);
+
+    expect(poolDataAfter).to.eql(expectedPoolData);
+    expect(userDataAfter).to.eql(expectedUserData);
+  }
```
Comment (qpzm) [inline test/utils/helpers.ts:134]: The definition lies here.

File: test/withdraw.test.ts [hunk 4]
```diff
@@ -347,74 +341,6 @@ describe('StakingPool.withdraw', () => {
         expect(poolDataAfterWithdraw).eql(expectedPoolDataWithdraw);
         expect(userDataAfterWithdraw).eql(expectedUserDataWithdraw);
       });
-
-
-      context('when admin sets bob as a manager', async () => {
-        it('bob becomes a manager and extend the pool and alice withdraw', async () => {
```
Comment (qpzm) [inline test/withdraw.test.ts:353]: I thought `setManager` didn't have to be tested in `stake.test.ts`, `withdraw.test.ts`, `claim.test.ts`, so extracted the tests in `stake.test.ts` into `extendPoolTest.ts`. 
If you think it is not enough, please comment.

File: test/close.test.ts [hunk 5]
```diff
@@ -40,17 +37,37 @@ describe('StakingPool.closePool', () => {
   beforeEach(async () => {
     testEnv = await loadFixture(fixture);
     actions = createTestActions(testEnv);
-    await actions.faucetAndApproveReward(deployer, MAX_UINT_AMOUNT);
+    await actions.faucetAndApproveReward(deployer, constants.MaxUint256.toString());
```
Comment (jaeboklee-modo) [inline test/close.test.ts]: `MaxUint256` doesn't need to be converted to string beacuse the type of input is BigNumberish.

Comment (jaeboklee-modo) [reply test/close.test.ts]: or How about making action functions receive both string type and bignumber type?

Comment (qpzm) [reply test/close.test.ts]: It seems better to change all helper functions to take BigNumberish `amount`.

Comment (jaeboklee-modo) [reply test/close.test.ts]: absolutely 👍

File: contracts/libraries/StakingPoolLogicV2.sol [hunk 6]
```diff
@@ -81,7 +78,7 @@ library StakingPoolLogicV2 {
     poolData.startTimestamp = roundStartTimestamp;
     poolData.endTimestamp = roundStartTimestamp + duration;
     poolData.lastUpdateTimestamp = roundStartTimestamp;
-    poolData.rewardIndex = 1e18;
+    // poolData.rewardIndex = 0;
```
Comment (donguks) [inline contracts/libraries/StakingPoolLogicV2.sol]: I think, we can remove the comment.

File: test/claim.test.ts [hunk 7]
```diff
@@ -20,20 +20,11 @@ describe('StakingPool.claim', () => {
...
-
-  const month_3 = BigNumber.from(8);
-  const day_3 = BigNumber.from(20);
-
-  const firstRoundStartTimestamp = toTimestamp(year, month_1, day_1, BigNumber.from(10));
-  const secondRoundStartTimestamp = toTimestamp(year, month_2, day_2, BigNumber.from(10));
-  const thirdRoundStartTimestamp = toTimestamp(year, month_3, day_3, BigNumber.from(10));
+  const firstRoundStartTimestamp = toTimestamp("2022.07.07 10:00:00Z")
+  const secondRoundStartTimestamp = toTimestamp("2022.07.08 10:00:00Z");
+  const thirdRoundStartTimestamp = toTimestamp("2022.08.20 10:00:00Z");
```
Comment (donguks) [inline test/claim.test.ts:27]: I like it!

File: test/close.test.ts [hunk 8]
```diff
@@ -40,17 +37,37 @@ describe('StakingPool.closePool', () => {
...
     });
 
+    it('reverts if a general account closes the pool', async () => {
+      await expect(testEnv.stakingPool.connect(alice).closePool()
+      ).to.be.revertedWith('Ownable: caller is not the owner');
+    });
+
+    it('reverts if open the pool already finished', async () => {
+      const rewardPersecond = BigNumber.from(utils.parseEther('1'));
+      const startAt = BigNumber.from(moment("2022.04.18 19:00:00", 'YYYY.MM.DD hh:mm:ss Z').unix())
```
Comment (donguks) [inline test/close.test.ts]: If startedAt is Bignumberish, we don't need to use `Bignumber.from`.
How about updating `initNewPoolAndTransfer`'s startedAt type?

File: test/close.test.ts [hunk 9]
```diff
@@ -40,17 +37,37 @@ describe('StakingPool.closePool', () => {
...
       await advanceTimeTo(startTimestamp);
     });
 
+    it('reverts if a general account closes the pool', async () => {
+      await expect(testEnv.stakingPool.connect(alice).closePool()
+      ).to.be.revertedWith('Ownable: caller is not the owner');
+    });
+
+    it('reverts if open the pool already finished', async () => {
+      const rewardPersecond = BigNumber.from(utils.parseEther('1'));
```
Comment (donguks) [inline test/close.test.ts]: `parseEther` already returns BigNubmer.

File: test/extendPool.test.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
+
```
Comment (donguks) [inline test/extendPool.test.ts]: Empty line!

File: test/extendPool.test.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+const { loadFixture } = waffle;
+
+describe('StakingPool.extendPool', () => {
+  let testEnv: TestEnv;
+  let actions: TestHelperActions;
+
+  const [deployer, alice, bob, carol] = waffle.provider.getWallets();
+
+  const rewardPersecond = BigNumber.from(utils.parseEther('1'));
```
Comment (donguks) [inline test/extendPool.test.ts]: We can remove `BigNumber.from` from `rewardPerSecond` and `newRewardPersecond`. Theses are already BigNumber.

**Comment by qpzm** [comment]
I guess what @jaeboklee-modo  mentioned in the scrum is the same as [elysia-dev/elyfi-v2@`3aa6669` (#40)](https://github.com/elysia-dev/elyfi-v2/pull/40/commits/3aa66693eb1520014d50234d61a288ddaf44e594).
If all users withdraw and the pool becomes empty, `getRewardIndex` should return the current index, not 0.
The initial value itself is not important but returning the current index is the key.  

**Reference**
https://github.com/elysia-dev/elyfi-stakingpool/commit/d4e69cfadf7d7331f46251d0c04d7a0d708fe18f#diff-a66ec54bef1103cd57502f4dc662e63f3ffd9429fc9e658b5733ebc3d56855ecR37-R38

File: test/extendPool.test.ts [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  const duration = BigNumber.from(30).mul(SECONDSPERDAY);
```
Comment (donguks) [inline test/extendPool.test.ts]: Duration can be BigNumberish and SECONDSPERDAY can be number.
We can remove `BigNumber.from`

File: test/extendPool.test.ts [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+import { waffle } from 'hardhat';
+import { TestEnv } from './types';
+import { RAY, SECONDSPERDAY } from './utils/constants';
+import { setTestEnv } from './utils/testEnv';
+import { createTestActions, TestHelperActions } from './utils/helpers';
+import { advanceTimeTo, resetTimestampTo, toTimestamp } from './utils/time';
+
+const { loadFixture } = waffle;
+
+describe('StakingPool.extendPool', () => {
```
Comment (donguks) [inline test/extendPool.test.ts:11]: Code splitting is always good.

File: test/stake.test.ts [hunk 14]
```diff
@@ -1,39 +1,40 @@
...
 describe('StakingPool.stake', () => {
   let testEnv: TestEnv;
   let actions: TestHelperActions;
+  let userDataBefore: UserData;
+  let poolDataBefore: PoolData;
+  let userDataAfter: UserData;
+  let poolDataAfter: PoolData;
+
+  const fetchDataBeforeAction = async (wallet: Wallet) => {
+    userDataBefore = await actions.getUserData(alice);
```
Comment (donguks) [inline test/stake.test.ts]: `alice -> wallet`

File: test/withdraw.test.ts (hunk 4 above)
Comment (donguks) [reply test/withdraw.test.ts:353]: I totally agree with you.

**Comment by jaeboklee-modo** [comment]
@qpzm Thanks for your check! I took a look at the [AAVE Distribution contract](https://github.com/aave/aave-stake-v2/blob/master/contracts/stake/AaveDistributionManager.sol), which uses the same index logic, and found that their initial index value is 0. 
Also, come to think of it, returning the current index in some situations is the key, not setting the initial value as 0. The below code shows what you mentioned. 
https://github.com/aave/aave-v3-periphery/blob/master/contracts/rewards/RewardsDistributor.sol#L484

File: test/extendPool.test.ts (hunk 11 above)
Comment (qpzm) [reply test/extendPool.test.ts]: Thanks! Fixed in [199846a](https://github.com/elysia-dev/stakingpool-v2/pull/26/commits/199846a53581169f2f345716062ffee40a716ef7).

File: test/extendPool.test.ts (hunk 12 above)
Comment (qpzm) [reply test/extendPool.test.ts]: Fixed in [199846a](https://github.com/elysia-dev/stakingpool-v2/pull/26/commits/199846a53581169f2f345716062ffee40a716ef7).

File: test/close.test.ts (hunk 8 above)
Comment (qpzm) [reply test/close.test.ts]: I agree. Fixed its type to `BigNumberish`.

File: test/stake.test.ts (hunk 14 above)
Comment (qpzm) [reply test/stake.test.ts]: Fixed in [e1df38c](https://github.com/elysia-dev/stakingpool-v2/pull/26/commits/e1df38c26c9bc987e9a2bbf6d7737c691f2ab657).

File: contracts/libraries/StakingPoolLogicV2.sol (hunk 6 above)
Comment (qpzm) [reply contracts/libraries/StakingPoolLogicV2.sol]: Fixed in [ecc8c28](https://github.com/elysia-dev/stakingpool-v2/pull/26/commits/ecc8c28d61a2497311b748aaeb3f765c9ee1def0).

**Pull request merged**